protocol/contracts/libraries/Silo/LibTokenSilo.sol

```diff
@@ -96,13 +96,25 @@ library LibTokenSilo {
             s.a[account].deposits[token][id].amount,
             s.a[account].deposits[token][id].bdv
         );
+
+        // If amount to remove is greater than the amount in the Deposit, migrate legacy Deposit to new Deposit
+        if (amount > crateAmount) {
+            // If Unripe Deposit, fetch whole Deposit balance and delete legacy deposit references.
+            if (LibUnripeSilo.isUnripeBean(token)) {
+                (crateAmount, crateBDV) = LibUnripeSilo.unripeBeanDeposit(account, id);
+                LibUnripeSilo.removeUnripeBeanDeposit(account, id);
+            } else if (LibUnripeSilo.isUnripeLP(token)) {
+                (crateAmount, crateBDV) = LibUnripeSilo.unripeLPDeposit(account, id);
+                LibUnripeSilo.removeUnripeLPDeposit(account, id);
+            }
+            require(crateAmount >= amount, "Silo: Crate balance too low.");
+        }
```

Comment:
Are these changes so that there's no slots storing "0" left on chain? Or what's the main idea behind doing it this way?

Reply:
Currently Beanstalk checks the Silo V2 reference and if there is not enough deposit, it then checks the legacy Silo V1 reference. There is a rounding error that can occur when performing a partial Withdrawal from a legacy Silo V1 storage reference. To prevent this from happening, this change removes the whole legacy Silo V1 reference (event in the case of a partial remove) and adds the remaining amount + BDV to the current Silo V2 storage reference.

By only performing full removes (deletes) from the legacy reference, it guarantees that the rounding error can't occur.

Reply:
Awesome, thanks for the clarity